src/Common/SensitiveDataMasker.cpp

```diff
@@ -93,22 +94,30 @@ class SensitiveDataMasker::MaskingRule
 
 SensitiveDataMasker::~SensitiveDataMasker() = default;
 
-std::unique_ptr<SensitiveDataMasker> SensitiveDataMasker::sensitive_data_masker = nullptr;
+std::shared_ptr<SensitiveDataMasker> SensitiveDataMasker::sensitive_data_masker = nullptr;
+std::mutex SensitiveDataMasker::instance_mutex;
 
-void SensitiveDataMasker::setInstance(std::unique_ptr<SensitiveDataMasker> sensitive_data_masker_)
+void SensitiveDataMasker::setInstance(std::shared_ptr<SensitiveDataMasker> sensitive_data_masker_)
 {
+
     if (!sensitive_data_masker_)
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Logical error: the 'sensitive_data_masker' is not set");
 
+    std::lock_guard lock(instance_mutex);
     if (sensitive_data_masker_->rulesCount() > 0)
     {
         sensitive_data_masker = std::move(sensitive_data_masker_);
     }
+    else
+    {
+        sensitive_data_masker.reset();
+    }
 }
 
-SensitiveDataMasker * SensitiveDataMasker::getInstance()
+std::shared_ptr<SensitiveDataMasker> SensitiveDataMasker::getInstance()
 {
-    return sensitive_data_masker.get();
+    std::lock_guard lock(instance_mutex);
```

Comment:
taking locks for every log line can hurt the performance. Maybe you can try using atomic pointers? (now we have c++20, see std::atomic<std::shared_ptr>)

Reply:
Thanks! I'll give it a try

Reply:
@filimonov thanks for the great hint! Unfortunately it seems that compilers don't support it yet :( 

https://en.cppreference.com/w/cpp/compiler_support/20 shows that clang with libc++ doesn't support it at all

Reply:
Indeed. Check also Common/MultiVersion.h 

Also i see some code inside clickhouse uses also atomic_shared_ptr from boost/smart_ptr

Reply:
Oh nice! It seems that `MultiVersion` is just something I needed here. Thanks!